# CALIDAD DE DATOS

## IMPORTAR PAQUETES

In [273]:
import numpy as np
import pandas as pd
from janitor import clean_names

#Automcompletar rápido
%config IPCompleter.greedy=True

import warnings
warnings.filterwarnings('ignore')

## IMPORTAR LOS DATOS

Sustituir la ruta del proyecto.

In [229]:
ruta_proyecto = 'C:/Users/pelop/OneDrive/Desktop/Curso Data Science Pedro/2 CURSO DATA SCIENCE/03_MACHINE_LEARNING/07_CASOS/01_LEADSCORING'

Sustituir el nombre_fichero_datos.

In [230]:
nombre_fichero_datos = 'trabajo.csv'

Cargar los datos.

In [231]:
ruta_completa = ruta_proyecto + '/02_Datos/03_Trabajo/' + nombre_fichero_datos

df = pd.read_csv(ruta_completa,index_col='id',sep=',')
df

origen          fuente no_enviar_email no_llamar  \
id                                                                          
660737                      API            Chat              No        No   
660728                      API  Organic Search              No        No   
660727  Landing Page Submission  Direct Traffic              No        No   
660719  Landing Page Submission  Direct Traffic              No        No   
660673  Landing Page Submission          Google              No        No   
...                         ...             ...             ...       ...   
579615  Landing Page Submission  Direct Traffic              No        No   
579564  Landing Page Submission  Direct Traffic             Yes        No   
579546  Landing Page Submission  Direct Traffic              No        No   
579545  Landing Page Submission  Direct Traffic             Yes        No   
579538  Landing Page Submission          Google              No        No   

        compra  visitas_total  tiempo_en_site_total  paginas_vistas_visita  \
id                                                                           
660737       0            0.0                     0                   0.00   
660728       0            5.0                   674                   2.50   
660727       1            2.0                  1532                   2.00   
660719       0            1.0                   305                   1.00   
660673       1            2.0                  1640                   2.00   
...        ...            ...                   ...                    ...   
579615       1            5.0                   210                   2.50   
579564       1            8.0                  1845                   2.67   
579546       0            2.0                   238                   2.00   
579545       0            2.0                   199                   2.00   
579538       1            3.0                   499                   3.00   

                  ult_actividad                     ambito   ocupacion  \
id                                                                       
660737  Page Visited on Website                     Select  Unemployed   
660728             Email Opened                     Select  Unemployed   
660727             Email Opened    Business Administration     Student   
660719              Unreachable      Media and Advertising  Unemployed   
660673             Email Opened    Supply Chain Management  Unemployed   
...                         ...                        ...         ...   
579615                 SMS Sent    Business Administration  Unemployed   
579564        Email Marked Spam     IT Projects Management  Unemployed   
579546                 SMS Sent      Media and Advertising  Unemployed   
579545                 SMS Sent    Business Administration  Unemployed   
579538                 SMS Sent  Human Resource Management         NaN   

       conociste_google conociste_revista conociste_periodico  \
id                                                              
660737               No                No                  No   
660728               No                No                  No   
660727               No                No                  No   
660719               No                No                  No   
660673               No                No                  No   
...                 ...               ...                 ...   
579615               No                No                  No   
579564               No                No                  No   
579546               No                No                  No   
579545               No                No                  No   
579538               No                No                  No   

       conociste_youtube conociste_facebook conociste_referencias  \
id                                                                  
660737                No                 No                  

In [232]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6365 entries, 660737 to 579538
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   origen                 6365 non-null   object 
 1   fuente                 6343 non-null   object 
 2   no_enviar_email        6365 non-null   object 
 3   no_llamar              6365 non-null   object 
 4   compra                 6365 non-null   int64  
 5   visitas_total          6268 non-null   float64
 6   tiempo_en_site_total   6365 non-null   int64  
 7   paginas_vistas_visita  6268 non-null   float64
 8   ult_actividad          6295 non-null   object 
 9   ambito                 5378 non-null   object 
 10  ocupacion              4492 non-null   object 
 11  conociste_google       6365 non-null   object 
 12  conociste_revista      6365 non-null   object 
 13  conociste_periodico    6365 non-null   object 
 14  conociste_youtube      6365 non-null   object 
 15  co

In [233]:
df.head()

origen          fuente no_enviar_email no_llamar  \
id                                                                          
660737                      API            Chat              No        No   
660728                      API  Organic Search              No        No   
660727  Landing Page Submission  Direct Traffic              No        No   
660719  Landing Page Submission  Direct Traffic              No        No   
660673  Landing Page Submission          Google              No        No   

        compra  visitas_total  tiempo_en_site_total  paginas_vistas_visita  \
id                                                                           
660737       0            0.0                     0                    0.0   
660728       0            5.0                   674                    2.5   
660727       1            2.0                  1532                    2.0   
660719       0            1.0                   305                    1.0   
660673       1            2.0                  1640                    2.0   

                  ult_actividad                   ambito   ocupacion  \
id                                                                     
660737  Page Visited on Website                   Select  Unemployed   
660728             Email Opened                   Select  Unemployed   
660727             Email Opened  Business Administration     Student   
660719              Unreachable    Media and Advertising  Unemployed   
660673             Email Opened  Supply Chain Management  Unemployed   

       conociste_google conociste_revista conociste_periodico  \
id                                                              
660737               No                No                  No   
660728               No                No                  No   
660727               No                No                  No   
660719               No                No                  No   
660673               No                No                  No   

       conociste_youtube conociste_facebook conociste_referencias  \
id                                                                  
660737                No                 No                    No   
660728                No                 No                    No   
660727                No                 No                    No   
660719                No                 No                    No   
660673                No                 No                    No   

        score_actividad  score_perfil descarga_lm  
id                                                 
660737             15.0          15.0          No  
660728             15.0          15.0          No  
660727             14.0          20.0         Yes  
660719             13.0          17.0          No  
660673             14.0          20.0          No

## CORRECCIÓN DE NOMBRES DE VARIABLES

No es necesaria en este caso

## VISIÓN GENERAL

In [234]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6365 entries, 660737 to 579538
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   origen                 6365 non-null   object 
 1   fuente                 6343 non-null   object 
 2   no_enviar_email        6365 non-null   object 
 3   no_llamar              6365 non-null   object 
 4   compra                 6365 non-null   int64  
 5   visitas_total          6268 non-null   float64
 6   tiempo_en_site_total   6365 non-null   int64  
 7   paginas_vistas_visita  6268 non-null   float64
 8   ult_actividad          6295 non-null   object 
 9   ambito                 5378 non-null   object 
 10  ocupacion              4492 non-null   object 
 11  conociste_google       6365 non-null   object 
 12  conociste_revista      6365 non-null   object 
 13  conociste_periodico    6365 non-null   object 
 14  conociste_youtube      6365 non-null   object 
 15  co

## TIPOS DE DATOS

### Identificación

In [235]:
df.dtypes

origen                    object
fuente                    object
no_enviar_email           object
no_llamar                 object
compra                     int64
visitas_total            float64
tiempo_en_site_total       int64
paginas_vistas_visita    float64
ult_actividad             object
ambito                    object
ocupacion                 object
conociste_google          object
conociste_revista         object
conociste_periodico       object
conociste_youtube         object
conociste_facebook        object
conociste_referencias     object
score_actividad          float64
score_perfil             float64
descarga_lm               object
dtype: object

### Corrección

In [236]:
df = df.astype({'visitas_total':'Int64'})

df.dtypes

origen                    object
fuente                    object
no_enviar_email           object
no_llamar                 object
compra                     int64
visitas_total              Int64
tiempo_en_site_total       int64
paginas_vistas_visita    float64
ult_actividad             object
ambito                    object
ocupacion                 object
conociste_google          object
conociste_revista         object
conociste_periodico       object
conociste_youtube         object
conociste_facebook        object
conociste_referencias     object
score_actividad          float64
score_perfil             float64
descarga_lm               object
dtype: object

## VALORES ÚNICOS

### Identificación

In [237]:
df.nunique().sort_values()

conociste_youtube           1
conociste_revista           1
descarga_lm                 2
conociste_referencias       2
no_enviar_email             2
no_llamar                   2
compra                      2
conociste_facebook          2
conociste_periodico         2
conociste_google            2
origen                      4
ocupacion                   6
score_perfil               10
score_actividad            12
ult_actividad              15
fuente                     16
ambito                     19
visitas_total              40
paginas_vistas_visita     100
tiempo_en_site_total     1572
dtype: int64

In [238]:
df.conociste_youtube.value_counts()

conociste_youtube
No    6365
Name: count, dtype: int64

### Corrección

In [239]:
df.drop(columns = ['conociste_youtube','conociste_revista'], inplace = True)
df.nunique().sort_values()

descarga_lm                 2
conociste_referencias       2
no_enviar_email             2
no_llamar                   2
compra                      2
conociste_facebook          2
conociste_periodico         2
conociste_google            2
origen                      4
ocupacion                   6
score_perfil               10
score_actividad            12
ult_actividad              15
fuente                     16
ambito                     19
visitas_total              40
paginas_vistas_visita     100
tiempo_en_site_total     1572
dtype: int64

## DUPLICADOS

### Identificación

In [240]:
df.duplicated().sum()

999

### Corrección

In [241]:
df.drop_duplicates(inplace = True)
df.duplicated().sum()

0

## SEPARAR NUMÉRICAS Y CATEGÓRICAS

### Categóricas

In [242]:
cat = df.select_dtypes(exclude = 'number').copy()

### Numéricas

In [243]:
num = df.select_dtypes(include='number').copy()

## GESTIÓN DE CATEGÓRICAS

### Nulos

#### Identificación

In [244]:
cat.isna().sum().sort_values(ascending = False)

ocupacion                1375
ambito                    489
ult_actividad              70
fuente                     14
origen                      0
no_enviar_email             0
no_llamar                   0
conociste_google            0
conociste_periodico         0
conociste_facebook          0
conociste_referencias       0
descarga_lm                 0
dtype: int64

#### Corrección

##### Eliminar la variable

En este caso la variable ocupación (que es la más susceptible a ser eliminada) es bastante importante. La dejamos.

##### Eliminar registros

*Cuidado: Si se eliminan registros hay que eliminarlos tanto de cat como de num*. En este caso no eliminamos ningún registro.

##### Imputar valores

In [245]:
cat.ocupacion.value_counts() # A la vista de los resultados, imputamos por la moda

ocupacion
Unemployed              3413
Working Professional     439
Student                  115
Other                     12
Businessman                6
Housewife                  6
Name: count, dtype: int64

In [246]:
cat.ambito.value_counts() # A la vista de los resultados, imputamos por la moda

ambito
Select                               864
Finance Management                   665
Human Resource Management            551
Marketing Management                 541
Operations Management                353
Business Administration              281
IT Projects Management               270
Supply Chain Management              244
Banking, Investment And Insurance    239
Media and Advertising                150
Travel and Tourism                   149
International Business               121
Healthcare Management                100
Hospitality Management                82
E-COMMERCE                            75
Retail Management                     72
Rural and Agribusiness                52
E-Business                            40
Services Excellence                   28
Name: count, dtype: int64

In [247]:
cat.ult_actividad.value_counts() # A la vista de los resultados, imputamos por valor desconocido

ult_actividad
Email Opened                    2024
SMS Sent                        1646
Page Visited on Website          450
Converted to Lead                303
Chat Conversation                291
Email Bounced                    205
Email Link Clicked               150
Form Submitted on Website         77
Unreachable                       71
Unsubscribed                      46
Had a Phone Conversation          18
View in browser link Clicked       6
Approached upfront                 5
Email Received                     2
Email Marked Spam                  2
Name: count, dtype: int64

In [248]:
cat.fuente.value_counts() # A la vista de los resultados, imputamos por valor desconocido (ya que las dos priemras fuentes son prácticamente igual de probables)

fuente
Google               1986
Direct Traffic       1797
Organic Search        838
Chat                  341
Reference             239
Referral Sites         96
Facebook               36
google                  5
bing                    5
Live Chat               2
Click2call              2
blog                    1
Pay per Click Ads       1
Social Media            1
youtubechannel          1
Press_Release           1
Name: count, dtype: int64

##### Imputar por un valor

###### Variables a imputar

Pensando en la fase de producción, vamos a imputar por valor no solo *ult_actividad* y *fuente* sino todas las demás que no se imputan por moda. Ya que aunque ahora no tengan nulos en un futuro podría aparecer alguno.

In [249]:
to_be_removed = {'ambito','ocupacion'}
var_imputar_valor = [item for item in cat.columns.to_list() if item not in to_be_removed]

var_imputar_valor

['origen',
 'fuente',
 'no_enviar_email',
 'no_llamar',
 'ult_actividad',
 'conociste_google',
 'conociste_periodico',
 'conociste_facebook',
 'conociste_referencias',
 'descarga_lm']

###### Valor por el que imputar

In [250]:
valor = 'DESCONOCIDO'

###### Imputar

In [251]:
cat[var_imputar_valor] = cat[var_imputar_valor].fillna(valor)

##### Imputar por la moda

###### Variables a imputar

In [252]:
var_imputar_moda = ['ocupacion','ambito']

###### Funcion para imputar con la moda

In [253]:
def imputar_moda(variable):
    return(variable.fillna(variable.mode()[0]))

###### Imputar con la moda

In [254]:
cat[var_imputar_moda] = cat[var_imputar_moda].apply(imputar_moda)

In [255]:
cat.isna().sum()

origen                   0
fuente                   0
no_enviar_email          0
no_llamar                0
ult_actividad            0
ambito                   0
ocupacion                0
conociste_google         0
conociste_periodico      0
conociste_facebook       0
conociste_referencias    0
descarga_lm              0
dtype: int64

### Atípicos

#### Categorías poco frecuentes

##### Función para agrupar categorías raras en 'OTROS'

In [256]:
def agrupar_cat_raras(variable, criterio = 0.02):
    #Calcula las frecuencias
    frecuencias = variable.value_counts(normalize=True)
    #Identifica las que están por debajo del criterio
    temp = [cada for cada in frecuencias.loc[frecuencias < criterio].index.values]
    #Las recodifica en 'OTROS'
    temp2 = np.where(variable.isin(temp),'OTROS',variable)
    #Devuelve el resultado
    return(temp2)

##### Variables a imputar

In [257]:
var_agrupar_cat_raras = cat.columns.to_list()
var_agrupar_cat_raras

['origen',
 'fuente',
 'no_enviar_email',
 'no_llamar',
 'ult_actividad',
 'ambito',
 'ocupacion',
 'conociste_google',
 'conociste_periodico',
 'conociste_facebook',
 'conociste_referencias',
 'descarga_lm']

##### Aplicar

In [258]:
for variable in var_agrupar_cat_raras:
    cat[variable] = agrupar_cat_raras(cat[variable],criterio = 0.02)

In [259]:
cat.fuente.value_counts()

fuente
Google            1986
Direct Traffic    1797
Organic Search     838
Chat               341
Reference          239
OTROS              165
Name: count, dtype: int64

## GESTIÓN DE NUMÉRICAS

### Nulos

#### Identificación

In [260]:
num.isna().sum().sort_values(ascending = False)

score_actividad          2348
score_perfil             2348
visitas_total              97
paginas_vistas_visita      97
compra                      0
tiempo_en_site_total        0
dtype: int64

#### Corrección

##### Eliminar la variable

Quizás sería más correcto eliminar los dos scorings, pues cerca del 40% de los valores son nulos. Sin embargo, como pueden ser variables bastante interesantes (si están bien construidas) vamos a dejarlas de momento.

##### Eliminar registros

*Cuidado: Si se eliminan registros hay que eliminarlos tanto de cat como de num*. No aplica ene ste caso.

##### Imputar valores

Para estas variables numéricas, y como norma, es mejor imputar por la mediana. Ya que no sabemos si la distribución es normal para poder utilizar la media. Nos curamos en salud.

##### Imputar por la mediana

###### Variables a imputar

Nota.- Vamos a imputar todas las variables numéricas por la mediana, pensando en que luego en la fase de producción podrían aparecernos nulos también en las categorías de *compra* y *tiempo_en_site_total*.

In [261]:
var_imputar_mediana = num.columns.to_list()

###### Funcion para imputar con la mediana

In [262]:
def imputar_mediana(variable):
    if pd.api.types.is_integer_dtype(variable):
        return(variable.fillna(int(variable.median())))
    else:
        return(variable.fillna(variable.median()))

###### Imputar con la mediana

In [263]:
num[var_imputar_mediana] = num[var_imputar_mediana].apply(imputar_mediana)

In [264]:
num.isna().sum()

compra                   0
visitas_total            0
tiempo_en_site_total     0
paginas_vistas_visita    0
score_actividad          0
score_perfil             0
dtype: int64

### Atípicos

#### Por desviación típica

##### Identificación

###### Definir el número de desviaciones típicas

In [265]:
num_desv_tip = 4

###### Función que devuelve los índices de los atípicos

In [266]:
def atipicos_desv_tip(variable, num_desv_tip = 4):
    #sacamos los nulos por ahora
    variable = variable.dropna()
    #calculamos los límites
    media = np.mean(variable)
    sd = np.std(variable)
    umbral = sd * num_desv_tip
    lim_inf = media - umbral
    lim_sup = media + umbral
    #encontramos los índices de los que están fuera de los límites
    indices = [indice for indice,valor in variable.items() if valor < lim_inf or valor > lim_sup]
    return(indices)

###### Función que cuenta el número de atípicos

In [267]:
def conteo_atipicos(df,variable, num_desv_tip = 4):
    atipicos = atipicos_desv_tip(df[variable], num_desv_tip)
    return(df.loc[atipicos,variable].value_counts())

###### Variables a aplicar

In [268]:
var_atipicos_dt = num.columns.to_list()

###### Aplicar

In [271]:
for variable in var_atipicos_dt:
    print('\n' + variable + ':\n')
    print(conteo_atipicos(num,variable,num_desv_tip).sort_index().sort_values(ascending = False))


compra:

Series([], Name: count, dtype: int64)

visitas_total:

visitas_total
27     5
28     1
30     1
32     1
41     1
42     1
43     1
54     1
55     1
74     1
115    1
141    1
251    1
Name: count, dtype: Int64

tiempo_en_site_total:

Series([], Name: count, dtype: int64)

paginas_vistas_visita:

paginas_vistas_visita
14.00    7
13.00    6
15.00    4
16.00    3
12.00    2
11.50    1
12.33    1
24.00    1
55.00    1
Name: count, dtype: int64

score_actividad:

score_actividad
10.0    42
9.0      4
8.0      3
18.0     3
7.0      1
Name: count, dtype: int64

score_perfil:

Series([], Name: count, dtype: int64)


A la vista de los resultados vamos a establecer dos cortes por winsorización manual para *visitas_total* > 50 y para *paginas_vistas_visita* > 20. La variable *score_actividad* la dejamos como está, pues no sabemos como está construida.

#### Por winsorización manual para una variable

##### Corrección

###### Aplicar

In [274]:
num['visitas_total'].clip(0,50,inplace=True)
num['paginas_vistas_visita'].clip(0,20,inplace=True)

In [275]:
for variable in var_atipicos_dt:
    print('\n' + variable + ':\n')
    print(conteo_atipicos(num,variable,num_desv_tip).sort_index().sort_values(ascending = False))


compra:

Series([], Name: count, dtype: int64)

visitas_total:

visitas_total
20    8
21    6
23    6
50    6
27    5
24    3
22    2
25    2
26    1
28    1
30    1
32    1
41    1
42    1
43    1
Name: count, dtype: Int64

tiempo_en_site_total:

Series([], Name: count, dtype: int64)

paginas_vistas_visita:

paginas_vistas_visita
11.00    13
14.00     7
13.00     6
15.00     4
16.00     3
12.00     2
20.00     2
11.50     1
12.33     1
Name: count, dtype: int64

score_actividad:

score_actividad
10.0    42
9.0      4
8.0      3
18.0     3
7.0      1
Name: count, dtype: int64

score_perfil:

Series([], Name: count, dtype: int64)


## GUARDAR DATASETS TRAS CALIDAD DE DATOS

Guarda los avances en df, cat y num con un sufijo para poder guardar o recuperar avances del proyecto.

En formato pickle para no perder las modificaciones de metadatos.

In [276]:
cat

origen          fuente no_enviar_email no_llamar  \
id                                                                          
660737                      API            Chat              No        No   
660728                      API  Organic Search              No        No   
660727  Landing Page Submission  Direct Traffic              No        No   
660719  Landing Page Submission  Direct Traffic              No        No   
660673  Landing Page Submission          Google              No        No   
...                         ...             ...             ...       ...   
579615  Landing Page Submission  Direct Traffic              No        No   
579564  Landing Page Submission  Direct Traffic             Yes        No   
579546  Landing Page Submission  Direct Traffic              No        No   
579545  Landing Page Submission  Direct Traffic             Yes        No   
579538  Landing Page Submission          Google              No        No   

                  ult_actividad                     ambito   ocupacion  \
id                                                                       
660737  Page Visited on Website                     Select  Unemployed   
660728             Email Opened                     Select  Unemployed   
660727             Email Opened    Business Administration     Student   
660719                    OTROS      Media and Advertising  Unemployed   
660673             Email Opened    Supply Chain Management  Unemployed   
...                         ...                        ...         ...   
579615                 SMS Sent    Business Administration  Unemployed   
579564                    OTROS     IT Projects Management  Unemployed   
579546                 SMS Sent      Media and Advertising  Unemployed   
579545                 SMS Sent    Business Administration  Unemployed   
579538                 SMS Sent  Human Resource Management  Unemployed   

       conociste_google conociste_periodico conociste_facebook  \
id                                                               
660737               No                  No                 No   
660728               No                  No                 No   
660727               No                  No                 No   
660719               No                  No                 No   
660673               No                  No                 No   
...                 ...                 ...                ...   
579615               No                  No                 No   
579564               No                  No                 No   
579546               No                  No                 No   
579545               No                  No                 No   
579538               No                  No                 No   

       conociste_referencias descarga_lm  
id                                        
660737                    No          No  
660728                    No          No  
660727                    No         Yes  
660719                    No          No  
660673                    No          No  
...                      ...         ...  
579615                    No          No  
579564                    No          No  
579546                    No         Yes  
579545                    No         Yes  
579538                    No          No  

[5366 rows x 12 columns]

In [277]:
num

compra  visitas_total  tiempo_en_site_total  paginas_vistas_visita  \
id                                                                           
660737       0              0                     0                   0.00   
660728       0              5                   674                   2.50   
660727       1              2                  1532                   2.00   
660719       0              1                   305                   1.00   
660673       1              2                  1640                   2.00   
...        ...            ...                   ...                    ...   
579615       1              5                   210                   2.50   
579564       1              8                  1845                   2.67   
579546       0              2                   238                   2.00   
579545       0              2                   199                   2.00   
579538       1              3                   499                   3.00   

        score_actividad  score_perfil  
id                                     
660737             15.0          15.0  
660728             15.0          15.0  
660727             14.0          20.0  
660719             13.0          17.0  
660673             14.0          20.0  
...                 ...           ...  
579615             14.0          20.0  
579564             15.0          17.0  
579546             14.0          19.0  
579545             13.0          20.0  
579538             15.0          16.0  

[5366 rows x 6 columns]

Vemos que tenemos 5366 registros en ambos datasets. Está todo correcto.

### Definir los nombres de los archivos

In [278]:
ruta_trabajo = ruta_proyecto + '/02_Datos/03_Trabajo/' + 'trabajo_resultado_calidad.pickle'
ruta_cat = ruta_proyecto + '/02_Datos/03_Trabajo/' + 'cat_resultado_calidad.pickle'
ruta_num = ruta_proyecto + '/02_Datos/03_Trabajo/' + 'num_resultado_calidad.pickle'

### Guardar los archivos

In [279]:
df.to_pickle(ruta_trabajo)
cat.to_pickle(ruta_cat)
num.to_pickle(ruta_num)